### DISCUSSION TITLE
How to implement custom responses for server slaves

### DISCUSSION THREAD

**Discussion by Uncharger** [pymodbus-dev/pymodbus#1213] [Help] [answered]
Hello,
I am currently working on a system simulator in python, which runs a modbus server with up to 32 slaves.
I need to implement responses to custom modbus functions.
I understand I need to create custom response/request classes from this article:
https://pymodbus.readthedocs.io/en/latest/source/example/custom_message.html

I however have no idea how to implement them into the sever context.
Help would be much appreciated. If further information is required I will happily provide.

**Comment by janiversen** [comment, marked as answer]
That is easy, just create a custom datastore (see [v2.5.3 example](https://github.com/riptideio/pymodbus/blob/dev/examples/v2.5.3/custom_datablock.py))
 that does what you want and pass it to the server with context=<your class>

the article you refer to is for clients not servers.

**Reply by Uncharger** [reply]
Ok, thank you. But I still am not sure how to use this information.
In my case, the client will send multiple different custom functions (0x20 and 0x21), which have multiple values that are not in the server's standard modbus map.
I am supposed to answer based on the slave's Serial Number, once responding with True/False answer for one function, and destroy and restore the slave with new modbus id for other function.

This is something I saw as an option in the client custom function page, but I really don't understand how to use it with these custom datablocks, which look to me like doing the same action every time after receiving a register value of their type?

**Comment by janiversen** [comment]
What do you mean "which have multiple values that are not in the server's standard modbus map" there are no standard modbus map for values.

Please have a closer look at the data block, validate_values and get/set _values are called with each incoming request, you get the function code, and you can respond as you please, with fixed values or calculated values.

You can add a datastore for each slave if you want (our server_async.py example actually does that).

Forget what you saw about the client, that have absolutely nothing to do with datastores.
